File: python/taflow/metrics/treynor_ratio.py
"""Benchmark-relative Treynor ratio metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import TreynorRatio as _Native
from ._input import as_paired_metric_series


class TreynorRatio:
    """Compute geometrically annualized excess return divided by market beta.

    TAFlow follows PerformanceAnalytics 2.1.0 ``TreynorRatio``: subtract the
    same per-period risk-free return from the aligned portfolio and benchmark
    returns, compute CAPM beta, geometrically annualize the portfolio's
    resulting excess-return observations, and divide that annualized return by
    beta. The source convention is pinned to the CRAN 2.1.0 tarball with
    SHA-256 ``fc801d39382818cd3a7052326b45d078302aef4d290c85dab83498ed4516d58d``.
    The executable independent oracle cross-check is QuantStats 0.0.81
    ``treynor_ratio`` on its mathematically equivalent zero-risk-free subset.
    TAFlow accepts an annual effective risk-free rate and converts it in Rust
    with ``expm1(log1p(rate) / periods_per_year)``.

    Warm-up requires two usable aligned pairs. Zero benchmark variance, zero
    beta, or a non-finite/non-real annualized excess return yields ``None``.
    Missing values are omitted pairwise under ``nan_policy="omit"`` or rejected
    by ``"raise"``; infinities and unequal lengths are rejected before native
    mutation. Inputs may be simple returns, log returns, positive equity
    levels, or non-cumulative period P&L. P&L requires separate positive initial
    capital for each stream. The first equity pair establishes conversion
    baselines and does not increment length. Mutating lifecycle methods are
    fluent; Rust owns conversion and allocation-free O(1) metric arithmetic.
    """

    def __init__(self) -> None:
        """Reject ambiguous construction; use a paired semantic factory."""
        raise TypeError(
            "use TreynorRatio.from_returns/from_equity/from_pnl/from_log_returns"
        )

    @classmethod
    def _create(
        cls,
        values: Any,
        benchmark_values: Any,
        input_mode: str,
        *,
        periods_per_year: float = 252.0,
        annual_risk_free_rate: float = 0.0,
        initial_equity: float | None = None,
        benchmark_initial_equity: float | None = None,
        nan_policy: str = "omit",
    ) -> "TreynorRatio":
        primary, benchmark = as_paired_metric_series(values, benchmark_values)
        state = cls.__new__(cls)
        state._state = _Native(
            input_mode,
            float(periods_per_year),
            float(annual_risk_free_rate),
            initial_equity,
            benchmark_initial_equity,
            nan_policy,
        )
        state._state.extend(primary, benchmark)
        return state

    @classmethod
    def from_returns(
        cls,
        returns: Any,
        benchmark_returns: Any,
        *,
        periods_per_year: float = 252.0,
        annual_risk_free_rate: float = 0.0,
        nan_policy: str = "omit",
    ) -> "TreynorRatio":
        """Construct from aligned chronological decimal simple returns."""
        return cls._create(
            returns,
            benchmark_returns,
            "returns",
            periods_per_year=periods_per_year,
            annual_risk_free_rate=annual_risk_free_rate,
            nan_policy=nan_policy,
        )

    @classmethod
    def from_log_returns(
        cls,
        log_returns: Any,
        benchmark_log_returns: Any,
        *,
        periods_per_year: float = 252.0,
        annual_risk_free_rate: float = 0.0,
        nan_policy: str = "omit",
    ) -> "TreynorRatio":
        """Construct from aligned chronological log returns converted by Rust."""
        return cls._create(
            log_returns,
            benchmark_log_returns,
            "log_returns",
            periods_per_year=periods_per_year,
            annual_risk_free_rate=annual_risk_free_rate,
            nan_policy=nan_policy,
        )

    @classmethod
    def from_equity(
        cls,
        equity: Any,
        benchmark_equity: Any,
        *,
        periods_per_year: float = 252.0,
        annual_risk_free_rate: float = 0.0,
        nan_policy: str = "omit",
    ) -> "TreynorRatio":
        """Construct from aligned positive equity or adjusted-price levels."""
        return cls._create(
            equity,
            benchmark_equity,
            "equity",
            periods_per_year=periods_per_year,
            annual_risk_free_rate=annual_risk_free_rate,
            nan_policy=nan_policy,
        )

    @classmethod
    def from_pnl(
        cls,
        pnl: Any,
        benchmark_pnl: Any,
        *,
        initial_equity: float,
        benchmark_initial_equity: float,
        periods_per_year: float = 252.0,
        annual_risk_free_rate: float = 0.0,
        nan_policy: str = "omit",
    ) -> "TreynorRatio":
        """Construct from aligned period P&L and separate initial capitals."""
        return cls._create(
            pnl,
            benchmark_pnl,
            "pnl",
            periods_per_year=periods_per_year,
            annual_risk_free_rate=annual_risk_free_rate,
            initial_equity=float(initial_equity),
            benchmark_initial_equity=float(benchmark_initial_equity),
            nan_policy=nan_policy,
        )

    def append(self, value: float, benchmark_value: float) -> "TreynorRatio":
        """Append one aligned pair in the selected domain and return this metric."""
        self._state.append(float(value), float(benchmark_value))
        return self

    def extend(self, values: Any, benchmark_values: Any) -> "TreynorRatio":
        """Append equal-length aligned series and return this metric."""
        primary, benchmark = as_paired_metric_series(values, benchmark_values)
        self._state.extend(primary, benchmark)
        return self

    @property
    def value(self) -> float | None:
        """Return the current Treynor ratio, or ``None`` until defined."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current scalar without replaying processed input."""
        return self._state.compute()

    def reset(self) -> "TreynorRatio":
        """Clear observations, preserve configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable aligned return pairs processed by Rust."""
        return len(self._state)


__all__ = ["TreynorRatio"]

File: crates/taflow-metrics/src/metrics/treynor_ratio.rs
use crate::{
    primitives::PairedMoments, MetricError, MetricInputKind, MetricResult, NanPolicy,
    PairedMetricInputState,
};

/// Geometrically annualized excess return divided by market beta.
#[derive(Debug, Clone)]
pub struct TreynorRatio {
    input: PairedMetricInputState,
    moments: PairedMoments,
    excess_growth_product: f64,
    periods_per_year: f64,
    period_risk_free_rate: f64,
}

impl TreynorRatio {
    /// Construct an empty state for aligned primary and benchmark input domains.
    pub fn new(
        primary_input_kind: MetricInputKind,
        benchmark_input_kind: MetricInputKind,
        periods_per_year: f64,
        annual_risk_free_rate: f64,
        nan_policy: NanPolicy,
    ) -> MetricResult<Self> {
        if !periods_per_year.is_finite() || periods_per_year <= 0.0 {
            return Err(MetricError::InvalidParameter {
                name: "periods_per_year",
                value: periods_per_year.to_string(),
                reason: "must be finite and greater than zero",
            });
        }
        if !annual_risk_free_rate.is_finite() || annual_risk_free_rate <= -1.0 {
            return Err(MetricError::InvalidParameter {
                name: "annual_risk_free_rate",
                value: annual_risk_free_rate.to_string(),
                reason: "must be finite and greater than -1",
            });
        }
        if matches!(
            primary_input_kind,
            MetricInputKind::RawPnl | MetricInputKind::Trades
        ) || matches!(
            benchmark_input_kind,
            MetricInputKind::RawPnl | MetricInputKind::Trades
        ) {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{primary_input_kind:?}/{benchmark_input_kind:?}"),
                reason: "Treynor ratio requires returns, log returns, equity, or period P&L with initial equity",
            });
        }

        let period_risk_free_rate = (annual_risk_free_rate.ln_1p() / periods_per_year).exp_m1();
        Ok(Self {
            input: PairedMetricInputState::new(
                primary_input_kind,
                benchmark_input_kind,
                nan_policy,
            )?,
            moments: PairedMoments::new(),
            excess_growth_product: 1.0,
            periods_per_year,
            period_risk_free_rate,
        })
    }

    /// Append one aligned pair and return the current Treynor ratio when defined.
    pub fn append(&mut self, primary: f64, benchmark: f64) -> MetricResult<Option<f64>> {
        if let Some((primary_return, benchmark_return)) = self.input.append(primary, benchmark)? {
            let primary_excess = primary_return - self.period_risk_free_rate;
            let benchmark_excess = benchmark_return - self.period_risk_free_rate;
            self.moments.append(primary_excess, benchmark_excess);
            self.excess_growth_product *= 1.0 + primary_excess;
        }
        Ok(self.value())
    }

    /// Append equal-length aligned slices through the same persistent state.
    pub fn extend(&mut self, primary: &[f64], benchmark: &[f64]) -> MetricResult<Option<f64>> {
        let input = &mut self.input;
        let moments = &mut self.moments;
        let product = &mut self.excess_growth_product;
        let risk_free = self.period_risk_free_rate;
        input.extend_slices(primary, benchmark, |primary_return, benchmark_return| {
            let primary_excess = primary_return - risk_free;
            let benchmark_excess = benchmark_return - risk_free;
            moments.append(primary_excess, benchmark_excess);
            *product *= 1.0 + primary_excess;
        })?;
        Ok(self.value())
    }

    /// Return the PerformanceAnalytics-compatible Treynor ratio.
    ///
    /// The numerator is `prod(1 + primary_return - period_risk_free_rate)
    /// ^ (periods_per_year / n) - 1`. Beta is the covariance of primary and
    /// benchmark excess returns divided by benchmark excess-return variance.
    /// At least two usable pairs, a non-zero finite beta, and a finite real
    /// annualized excess return are required.
    pub fn value(&self) -> Option<f64> {
        let benchmark_variance = self.moments.benchmark_variance(1)?;
        if benchmark_variance == 0.0 {
            return None;
        }
        let beta = self.moments.covariance(1)? / benchmark_variance;
        if beta == 0.0 || !beta.is_finite() {
            return None;
        }
        let exponent = self.periods_per_year / self.input.len() as f64;
        let annualized_excess = self.excess_growth_product.powf(exponent) - 1.0;
        let ratio = annualized_excess / beta;
        ratio.is_finite().then_some(ratio)
    }

    /// Return the current result without replaying prior observations.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving input and annual settings.
    pub fn reset(&mut self) {
        self.input.reset();
        self.moments.reset();
        self.excess_growth_product = 1.0;
    }

    /// Return the number of usable aligned return pairs processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable aligned return pairs have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}
